arch/lkl/kernel/syscalls.c

```diff
@@ -33,400 +32,125 @@ syscall_handler_t syscall_table[__NR_syscalls] = {
+				params[4], params[5]);
 
 	task_work_run();
 
 	return ret;
 }
 
-static irqreturn_t syscall_irq_handler(int irq, void *dev_id)
-{
-	struct syscall_thread_data *data = (struct syscall_thread_data *)dev_id;
-
-	wake_up(&data->wqh);
 
-	return IRQ_HANDLED;
-}
+#define CLONE_FLAGS (CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_THREAD |	\
+		     CLONE_SIGHAND | SIGCHLD)
 
-static void cleanup_syscall_threads(void);
+static int host_task_id;
+static struct task_struct *host0;
 
-int syscall_thread(void *_data)
+static int new_host_task(struct task_struct **task)
 {
-	struct syscall_thread_data *data;
-	struct syscall *s;
-	int ret;
-	static int count;
-
-	data = (struct syscall_thread_data *)_data;
-	init_waitqueue_head(&data->wqh);
-	list_add(&data->list, &syscall_threads);
-	init_completion(&data->stopped);
-
-	snprintf(current->comm, sizeof(current->comm), "ksyscalld%d", count++);
-
-	data->irq = lkl_get_free_irq("syscall");
-	if (data->irq < 0) {
-		pr_err("lkl: %s: failed to allocate irq: %d\n", __func__,
-		       data->irq);
-		return data->irq;
-	}
-
-	ret = request_irq(data->irq, syscall_irq_handler, 0, current->comm,
-			  data);
-	if (ret) {
-		pr_err("lkl: %s: failed to request irq %d: %d\n", __func__,
-		       data->irq, ret);
-		lkl_put_irq(data->irq, "syscall");
-		data->irq = -1;
-		return ret;
-	}
-
-	pr_info("lkl: syscall thread %s initialized (irq%d)\n", current->comm,
-		data->irq);
-
-	/* system call thread is ready */
-	lkl_ops->sem_up(data->completion);
-
-	while (1) {
-		wait_event(data->wqh,
-			   (s = dequeue_syscall(data)) != NULL || data->stop);
-
-		if (data->stop || s->no == __NR_reboot)
-			break;
-
-		run_syscall(s);
-
-		lkl_ops->sem_up(data->completion);
-	}
+	pid_t pid;
 
-	if (data == &default_syscall_thread_data)
-		cleanup_syscall_threads();
+	switch_to_host_task(host0);
 
-	pr_info("lkl: exiting syscall thread %s\n", current->comm);
+	pid = kernel_thread(NULL, NULL, CLONE_FLAGS);
+	if (pid < 0)
+		return pid;
 
-	list_del(&data->list);
+	rcu_read_lock();
+	*task = find_task_by_pid_ns(pid, &init_pid_ns);
+	rcu_read_unlock();
 
-	free_irq(data->irq, data);
-	lkl_put_irq(data->irq, "syscall");
+	host_task_id++;
 
-	if (data->stop) {
-		complete(&data->stopped);
-	} else {
-		s->ret = 0;
-		lkl_ops->sem_up(data->completion);
-	}
+	snprintf((*task)->comm, sizeof((*task)->comm), "host%d", host_task_id);
 
 	return 0;
 }
 
-static unsigned int syscall_thread_data_key;
-
-static int syscall_thread_data_init(struct syscall_thread_data *data,
-				    void *completion)
+static void del_host_task(void *arg)
 {
-	data->mutex = lkl_ops->sem_alloc(1);
-	if (!data->mutex)
-		return -ENOMEM;
+	struct task_struct *task = (struct task_struct *)arg;
 
-	if (!completion)
-		data->completion = lkl_ops->sem_alloc(0);
-	else
-		data->completion = completion;
-	if (!data->completion) {
-		lkl_ops->sem_free(data->mutex);
-		data->mutex = NULL;
-		return -ENOMEM;
-	}
+	lkl_cpu_get();
```

Comment:
Should the return value be checked so the following can be skipped after shutdown is done?